### PR TITLE
Redirect to relative subpath for approval, relative parent path on success

### PR DESCRIPTION
**Pull request by liggitt**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1434983

* Moves grant approval endpoint from `/oauth/approve` to `/oauth/authorize/approve`
* Changes the redirect to the grant approval flow to be a relative path redirect (`Location: authorize/approve`)
* Changes the grant approval form action to be relative (`action="approve"`)
* Changes the redirect from the approval flow back to the authorize endpoint to be relative (`Location: ../authorize`)
* Adds validation to RequestHeaderIdentityProvider loginURL to catch two scenarios that would prevent relative redirects from working (trailing `/` and the proxy path not ending in `/authorize`)

TODO:
* [x] Test:
  * [x] Firefox
  * [x] IE 11
  * [x] IE 11 Intranet mode
  * [x] Safari
  * [x] Chrome
  * [x] iOS
* [x] Update unit/integration tests
* [ ] Update documentation
  * [ ] the auth proxy endpoint that proxies to `/oauth/authorize` must also proxy subpaths (notably `/oauth/authorize/approve`)
  * [ ] the auth proxy endpoint that proxies to `/oauth/authorize` must end with `.../authorize` (with no trailing slash)

### CODE REVIEW

File: pkg/auth/oauth/handlers/grant.go [hunk 1]
```diff
@@ -140,32 +141,41 @@ func (g *autoGrant) GrantNeeded(user user.Info, grant *api.Grant, w http.Respons
...
-		"redirect_uri": {grant.RedirectURI},
-	}.Encode()
-	http.Redirect(w, req, redirectURL.String(), http.StatusFound)
+	_, lastSegment := path.Split(req.URL.Path)
+
+	// Make our "then" URL a relative backstep
+	reqURL := &(*req.URL)
+	reqURL.Host = ""
+	reqURL.Scheme = ""
+	reqURL.Path = path.Join("..", lastSegment)
```
Comment (deads2k) [inline pkg/auth/oauth/handlers/grant.go]: Why go up a level and back down to the same level?

File: pkg/cmd/server/api/validation/oauth.go [hunk 2]
```diff
@@ -270,14 +271,26 @@ func ValidateRequestHeaderIdentityProvider(provider *api.RequestHeaderIdentityPr
...
+			if !strings.Contains(url.RawQuery, redirector.URLToken) && !strings.Contains(url.RawQuery, redirector.QueryToken) {
+				validationResults.AddWarnings(
+					field.Invalid(
+						fieldPath.Child("provider", "loginURL"),
+						provider.LoginURL,
+						fmt.Sprintf("query does not include %q or %q, redirect will not preserve original authorize parameters", redirector.URLToken, redirector.QueryToken),
+					),
+				)
+			}
+			if strings.HasSuffix(url.Path, "/") {
```
Comment (deads2k) [inline pkg/cmd/server/api/validation/oauth.go]: These didn't work before either?

File: pkg/auth/oauth/handlers/grant.go (hunk 1 above)
Comment (liggitt) [inline pkg/auth/oauth/handlers/grant.go]: gets rid of trailing slash

File: pkg/cmd/server/api/validation/oauth.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/oauth.go]: they didn't work for grant approval, they could have worked for authorization

File: pkg/auth/oauth/handlers/grant.go (hunk 1 above)
Comment (enj) [inline pkg/auth/oauth/handlers/grant.go]: Use `path.Base` or `strings.TrimRight`?  Regardless, probably needs a comment.

Comment (liggitt) [inline pkg/auth/oauth/handlers/grant.go]: you misunderstand, it gets rid of the trailing slash in the browser... if you just redirect to `.`, you'd get `/oauth/authorize/`

**Comment by liggitt** [comment]
[test]

File: pkg/auth/oauth/handlers/grant.go (hunk 1 above)
Comment (liggitt) [inline pkg/auth/oauth/handlers/grant.go]: commented

**Comment by enj** [comment]
LGTM 👍 :shipit:

**Comment by liggitt** [comment]
[merge]

**Comment by liggitt** [comment]
[test]

**Pull request merged**